File: app/src/main/java/com/erotc/learning/fragment/InitializeFragment.java
package com.erotc.learning.fragment;


import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.erotc.learning.R;
import com.erotc.learning.dao.DictionaryEntry;
import com.erotc.learning.dao.QuestionSet;
import com.erotc.learning.parser.DictionaryParser;
import com.erotc.learning.parser.QuestionParser;
import com.erotc.learning.repository.DictionaryRepository;

import org.json.JSONException;

import java.io.IOException;
import java.sql.SQLException;
import java.util.ArrayList;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link InitializeFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class InitializeFragment extends Fragment {
    private static final String LOG_TAG = InitializeFragment.class.getSimpleName();

    private TextView mLabel;

    public interface InitializeListener {
        void onDone();
    }

    private InitializeListener listener;

    public InitializeFragment() {
        // Required empty public constructor
    }

    public static InitializeFragment newInstance() {
        return new InitializeFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_initialize, container, false);

        mLabel = view.findViewById(R.id.label);

        initialize();

        return view;
    }

    @SuppressLint("StaticFieldLeak")
    private void initialize(){
        new AsyncTask<Void, String, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                try {
                    DictionaryRepository repository = DictionaryRepository.getInstance(getContext());

                    if ( repository.isDictionaryEmpty() ){
                        publishProgress(getString(R.string.message_preparing_dictionary));

                        DictionaryParser dictionaryParser = new DictionaryParser (
                                getResources().openRawResource(R.raw.dictionary)
                        );

                        ArrayList<DictionaryEntry> entryList = dictionaryParser.parse();
                        repository.saveDictionaryEntries(entryList);
                    }

                    if ( repository.isQuestionSetEmpty() ){
                        publishProgress(getString(R.string.message_preparing_game));

                        QuestionParser questionParser = new QuestionParser(
                                getResources().openRawResource(R.raw.questionset)
                        );

                        ArrayList<QuestionSet> questionSetList = questionParser.parse();
                        repository.saveQuestionSets(questionSetList);
                    }
                } catch (SQLException | IOException | JSONException e) {
                    e.printStackTrace();
                    Log.e(LOG_TAG, Log.getStackTraceString(e));
                }

                return null;
            }

            @Override
            protected void onProgressUpdate(String... values) {
                mLabel.setText(values[0]);
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                if ( listener != null ){
                    listener.onDone();
                }
            }
        }.execute();
    }

    public void setListener(InitializeListener listener) {
        this.listener = listener;
    }
}

File: app/src/main/java/com/erotc/learning/fragment/InitializeFragment.kt
package com.erotc.learning.fragment

import android.annotation.SuppressLint
import android.os.AsyncTask
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.erotc.learning.R
import com.erotc.learning.data.Assessment
import com.erotc.learning.data.Lecture
import com.erotc.learning.data.Topic
import com.erotc.learning.data.Tutorial
import com.erotc.learning.repository.LearnRepository
import com.erotc.learning.util.ApplicationUtil
import com.google.gson.reflect.TypeToken
import kotlinx.android.synthetic.main.fragment_initialize.*


/**
 * A simple [Fragment] subclass.
 * Use the [InitializeFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class InitializeFragment : Fragment() {
    interface InitializeListener {
        fun onDone()
    }

    private var listener: InitializeListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_initialize, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initialize()
    }

    @SuppressLint("StaticFieldLeak")
    private fun initialize() {
        val context = context ?: return

        object : AsyncTask<Void?, String?, Void?>() {
            override fun doInBackground(vararg p0: Void?): Void? {
                val repository = LearnRepository.getInstance(context)
                if (repository.isTopicEmpty) {
                    publishProgress(getString(R.string.message_preparing_topics))

                    val topicString = ApplicationUtil.getStringFromAsset(context, ApplicationUtil.FILE_TOPIC)
                    val topicListType = object : TypeToken<List<Topic>>(){}.type

                    repository.saveTopics(ApplicationUtil.gson.fromJson(topicString, topicListType))
                }

                if (repository.isLectureEmpty) {
                    publishProgress(getString(R.string.message_preparing_lectures))

                    val lectureString = ApplicationUtil.getStringFromAsset(context, ApplicationUtil.FILE_LECTURE)
                    val lectureListType = object : TypeToken<List<Lecture>>(){}.type

                    repository.saveLectures(ApplicationUtil.gson.fromJson(lectureString, lectureListType))
                }

                if (repository.isTutorialEmpty) {
                    publishProgress(getString(R.string.message_preparing_tutorials))

                    val tutorialString = ApplicationUtil.getStringFromAsset(context, ApplicationUtil.FILE_TUTORIAL)
                    val tutorialListType = object : TypeToken<List<Tutorial>>(){}.type

                    repository.saveTutorials(ApplicationUtil.gson.fromJson(tutorialString, tutorialListType))
                }

                if (repository.isAssessmentEmpty) {
                    publishProgress(getString(R.string.message_preparing_assessment))

                    val assessmentString = ApplicationUtil.getStringFromAsset(context, ApplicationUtil.FILE_ASSESSMENT)
                    val assessmentListType = object : TypeToken<List<Assessment>>(){}.type

                    repository.saveAssessments(ApplicationUtil.gson.fromJson(assessmentString, assessmentListType))
                }

                return null
            }

            override fun onProgressUpdate(vararg values: String?) {
                label.text = values[0]
            }

            override fun onPostExecute(aVoid: Void?) {
                listener?.onDone()
            }
        }.execute()
    }

    fun setListener(listener: InitializeListener?) {
        this.listener = listener
    }

    companion object {
        private val LOG_TAG = InitializeFragment::class.java.simpleName
        fun newInstance(): InitializeFragment {
            return InitializeFragment()
        }
    }
}